### PR TITLE
Push notifications v2

### PR DESCRIPTION
**Pull request by jigar288**
Project now has PWA push notifications using service workers. Users will be notified when the classes they're tracking have seats available.

### CODE REVIEW

File: api/src/routes/push-service/index.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+//!FIXME: dev purposes --> shouldn't be its own route
+//* TODO: for dev purposes have a logged in user trigger this route and send push notifications to all their subscriptionObjects
+router.post('/send-notifications', isAuthenticated, async (req: Request, res: Response) => {
+    // * Grab user id from session
+    const _id = (req.user as UserObject)._id
+
+    try {
+        //! FIXME: use env variables here
+        webpush.setGCMAPIKey('565395438650')
```
Comment (bmiddha) [inline api/src/routes/push-service/index.ts]: Clean hard coded values

File: api/src/routes/push-service/index.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            'BK_0D9VS_RrjJh3BRbdBifq6Ump45KpzfwWxk6P6sVOSTcrc89TzWlgtM1f7R7hOiKQsOxZHlGNGRiex02n9-9g',
+            'xRRruVND4fgEeBQoa3mld2ulOwXZxLtWAlaUyPuycpg'
+        )
+
+        // * grab pushSubscription objects from database & send them to client --> check console
+
+        const user = await User.findOne({ _id })
+
+        user.subscriptionObjects.forEach(element => {
+            webpush.sendNotification(element, 'CRN:' + '44033' + ' ' + 'OPEN')
```
Comment (bmiddha) [inline api/src/routes/push-service/index.ts:73]: clean hard coded values

File: api/src/util/notifier.ts [hunk 3]
```diff
@@ -42,4 +45,25 @@ export const notifyUser = async (user: UserObject, classData: any) => {
...
+    if (user.pushNotificationsEnabled) {
+        //! FIXME: use env variables here
+        // * GCMAPIKey is a cloud messaging id from google cloud console or firebase to help deliver the message
+        // * GCMAPIKey must also be declared in manifest.json file as "gcm_sender_id"
+        webpush.setGCMAPIKey('565395438650')
+
+        //* 2nd and 3rd arugment are keys generated once by web push only once
+        //* 2nd arugment: public key & 3rd arugment: private key for server
+        webpush.setVapidDetails(
+            'mailto:jigar@novusclub.org',
```
Comment (bmiddha) [inline api/src/util/notifier.ts]: clean hard coded values

File: client/package.json [hunk 4]
```diff
@@ -71,8 +73,7 @@
         "redux-thunk": "^2.3.0",
         "regenerator-runtime": "^0.13.5",
         "tiny-warning": "^1.0.3",
-        "react-select": "~3.1.0",
-        "@types/react-select": "~3.0.11"
+        "prettier": "^2.0.5"
```
Comment (bmiddha) [inline client/package.json]: should be a dev dependency

File: client/package.json [hunk 5]
```diff
@@ -51,6 +51,7 @@
         "@fortawesome/react-fontawesome": "^0.1.9",
         "@testing-library/jest-dom": "^5.5.0",
         "@testing-library/user-event": "^10.0.1",
+        "@types/react-select": "~3.0.11",
```
Comment (bmiddha) [inline client/package.json]: should be a dev dependency

File: client/public/manifest.json [hunk 6]
```diff
@@ -24,5 +24,6 @@
     "start_url": ".",
     "display": "fullscreen",
     "theme_color": "#000000",
-    "background_color": "#ffffff"
+    "background_color": "#ffffff",
+    "gcm_sender_id": "565395438650"
```
Comment (bmiddha) [inline client/public/manifest.json:28]: does this need to be hard coded here?

File: client/public/service-worker.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+//TODO: add active, install, and push eventListeners
```
Comment (bmiddha) [inline client/public/service-worker.js]: resolve todo comments

File: client/src/index.tsx [hunk 8]
```diff
@@ -5,7 +5,7 @@ import { App } from './App'
 
 import 'react-app-polyfill/ie11'
 import 'bootstrap/dist/css/bootstrap.css'
-import * as serviceWorker from './serviceWorker'
+// import * as serviceWorker from './serviceWorker'
```
Comment (bmiddha) [inline client/src/index.tsx]: remove the commented out service worker import and register

File: client/src/serviceWorker.ts [hunk 9]
```diff
@@ -10,6 +10,122 @@
 // To learn more about the benefits of this model and instructions on how to
 // opt-in, read https://bit.ly/CRA-PWA
 
+const applicationServerPublicKey =
+    'BK_0D9VS_RrjJh3BRbdBifq6Ump45KpzfwWxk6P6sVOSTcrc89TzWlgtM1f7R7hOiKQsOxZHlGNGRiex02n9-9g'
```
Comment (bmiddha) [inline client/src/serviceWorker.ts:15]: hard coded key

**Review by bmiddha — CHANGES_REQUESTED** [review]
Please remove the hard coded values from files.

File: api/src/routes/auth/index.ts [hunk 10]
```diff
@@ -38,8 +38,11 @@ router.get('/', isAuthenticated, async (req: Request, res: Response) => {
 router.post(
     '/login',
     passport.authenticate('local', { failureFlash: true }),
-    async (req: Request, res: Response) =>
+    async (req: Request, res: Response) => {
+        const user = User.findOne({ email: req.body.email })
+
         res.status(200).json(stripData(await User.findOne({ email: req.body.email })))
```
Comment (anarmawala) [inline api/src/routes/auth/index.ts:42]: Change this line to not call `User.findOne` twice

File: api/src/routes/push-service/index.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    // * Grab user id from session
+    const _id = (req.user as UserObject)._id
+
+    try {
+        //! FIXME: use env variables here
+        webpush.setGCMAPIKey('565395438650')
+        webpush.setVapidDetails(
+            'mailto:jigar@novusclub.org',
+            'BK_0D9VS_RrjJh3BRbdBifq6Ump45KpzfwWxk6P6sVOSTcrc89TzWlgtM1f7R7hOiKQsOxZHlGNGRiex02n9-9g',
+            'xRRruVND4fgEeBQoa3mld2ulOwXZxLtWAlaUyPuycpg'
```
Comment (anarmawala) [inline api/src/routes/push-service/index.ts]: These values should go into the ENV file

File: api/src/routes/push-service/index.ts (hunk 2 above)
Comment (anarmawala) [inline api/src/routes/push-service/index.ts:73]: Does this verify that the class is actually open?

File: client/public/service-worker.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+// * event handler for push notifications
+// TODO: this should only execute if user is logged in & tracking a class
```
Comment (anarmawala) [inline client/public/service-worker.js]: Resolve TODO

File: client/src/serviceWorker.ts [hunk 13]
```diff
@@ -10,6 +10,122 @@
...
+
+            return response.json()
+        })
+        .then(function (responseData) {
+            if (!(responseData.data && responseData.data.success)) {
+                throw new Error('Bad response from server.')
+            }
+        })
+
+    // TODO: write some code that enables and disables button on the front end that subscribes user for push notifications
```
Comment (anarmawala) [inline client/src/serviceWorker.ts]: Resolve TODO

File: client/src/serviceWorker.ts [hunk 14]
```diff
@@ -95,8 +215,43 @@ function checkValidServiceWorker(swUrl: string, config?: Config) {
...
+        }
+    } else {
+        console.log('statement not true')
+    }
+}
+
+//! FIXME: need to get rid of production statement
 export function register(config?: Config) {
-    if (process.env.NODE_ENV === 'production' && 'serviceWorker' in navigator) {
+    if ('serviceWorker' in navigator) {
```
Comment (anarmawala) [inline client/src/serviceWorker.ts]: This should be changed to **`'serviceWorker' in navigator && (process.env.NODE_ENV === production || isLocalhost)`**

File: client/src/serviceWorker.ts [hunk 15]
```diff
@@ -107,9 +262,13 @@ export function register(config?: Config) {
         }
 
         window.addEventListener('load', () => {
-            const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`
+            let swUrl = `${process.env.PUBLIC_URL}/service-worker.js`
```
Comment (anarmawala) [inline client/src/serviceWorker.ts]: We can remove `${process.env.PUBLIC_URL}` completely

**Review by anarmawala — CHANGES_REQUESTED** [review]
Good job 🌮! Left some comments

**Review by alexchomiak — APPROVED** [review]
Very well documented code, I see I'm starting to get to you Jigar 😉 As Bharat says I think it's adamant you remove the hardcoded variables and replace them with env variables that can be specified by the programmer hosting the application. Great work!

File: client/public/manifest.json (hunk 6 above)
Comment (jigar288) [reply client/public/manifest.json:28]: doesn't need to be a hardcoded value and can't use .env variables in manifest.json public file

File: api/src/routes/push-service/index.ts (hunk 2 above)
Comment (jigar288) [reply api/src/routes/push-service/index.ts:73]: this route isn't actually used. Its for our testing purposes

File: client/src/serviceWorker.ts [hunk 16]
```diff
@@ -231,12 +106,11 @@ function finishRegisteration(swUrl: string) {
 
 export function registerWithLogin() {
     if ('serviceWorker' in navigator) {
-        console.log('Service Worker is supported')
+        console.log('serviceWorker' in navigator && (process.env.NODE_ENV === 'production' || isLocalhost))
 
-        let swUrl = `${process.env.PUBLIC_URL}/service-worker.js`
+        let swUrl = `./service-worker.js`
```
Comment (anarmawala) [inline client/src/serviceWorker.ts]: Change of `const`

File: client/src/serviceWorker.ts [hunk 17]
```diff
@@ -231,12 +106,11 @@ function finishRegisteration(swUrl: string) {
 
 export function registerWithLogin() {
     if ('serviceWorker' in navigator) {
```
Comment (anarmawala) [inline client/src/serviceWorker.ts]: Change `'serviceWorker' in navigator` to `'serviceWorker' in navigator && (process.env.NODE_ENV === 'production' || isLocalhost)`

**Review by anarmawala — CHANGES_REQUESTED** [review]
Two things, I noticed but otherwise the code looks great. Good Job 🌮 🎉

**Review by bmiddha — APPROVED** [review]
Great job. 🌮

**Review by anarmawala — APPROVED** [review]
🌮

**Pull request merged**